### ISSUE TITLE
Manage upstream externals

### ISSUE THREAD

**Issue by johnd0e** [IITC-CE/ingress-intel-total-conversion#70]
Currently we have separate sub-directory with all upstream 'externals' (+ separate `images` subdirectory)

Most of them described in [ATTRIBUTION.md](https://github.com/IITC-CE/ingress-intel-total-conversion/blob/master/ATTRIBUTION.md).
Some of them have version info inside, but most of them not.
Some of them are even modified for iitc special needs, but in order to know details we have to dig though unclear commit history.

It's a pain to maintain and update these sources.

So I propose:
- [x] Update all 'externals' to current version

Then:
- [x] To document somewhere (for each 'external'):
  * version
  * url to upstream
  * add some comments
     + purpose, and where in iitc it is used
     + why was this component chosen
- [x] When updating write verbose commit message, incl. url / version info / hash.
- [x] If the source was modified, then we should:
  * add original unmodified
  * apply modification(s) as a separate commit(s), with verbose messages explaining purpose of changes
  * in case of heavy modification: keep separate branch, or make a fork  of original repo (could be hosted in https://github.com/IITC-CE/).
- [x] Find a way to avoid source modification
  * extend the component in some official way (if possible)
  * monkey-patch at run-time
  * push changes to upstream (via PR)
And in general:
- [x] Avoid in IITC/plugins own code inline including of 3rd-party code (and images).
Use `@@INCLUDExxx`-macros instead.

Related links:
- Remove customizations of third-party code library [#1165](https://github.com/iitc-project/ingress-intel-total-conversion/issues/1165)
- Which Directory for Reusable Code? [#1185](https://github.com/iitc-project/ingress-intel-total-conversion/issues/1185)

**Comment by johnd0e** [comment]
> If the source was modified, then we should:

May be this workflow could be enough:
* Create branch like `external-leaflet-update`.
* Add unmodified source files.
* Commit needed changes separately, possibly in several commits (in order to make modification as clear as possible)
* Merge into `master` (feature-branch could be keeped)

**Comment by johnd0e** [comment]
Here will be the list of all externals, where we should specify:
- current version (clear checkbox if it's below upstream)
- bold if the source was modified from upstream (specify exact mods)

(WIP)

### jQuery-related components:
- [ ] [**jQuery**](https://jquery.com/) (jquery-3.3.1.min.js) © [_JS Foundation_](https://js.foundation/): [[MIT](https://jquery.org/license/)]
- [**jQuery UI**](https://jqueryui.com/) (jquery-ui-1.12.1.min.js) © [_JS Foundation_](https://js.foundation/):[[MIT](https://jquery.org/license/)]
- [ ] [2b253c5] [**jquery-qrcode**](https://github.com/jeromeetienne/jquery-qrcode) (jquery.qrcode.min.js) © [_Jerome Etienne_](http://jetienne.com): [[MIT](https://github.com/jeromeetienne/jquery-qrcode/blob/master/MIT-LICENSE.txt)]
- [x] [74613a8 mod] [**jquery-taphold**](https://github.com/richadams/jquery-taphold) (taphold.js) © [_Rich Adams_](https://richadams.me/): [[CC BY 3.0](https://github.com/richadams/jquery-taphold/blob/master/LICENSE)]
  Mod: `touchcancel` support (added later to upstream too)
  jQuery: fixed deprecated functions (https://github.com/IITC-CE/jquery-taphold/commit/a85050a7744d780809e8f007f6dc43065cdece9c)
- [ ] [~1.2.0 Mod] [**Spectrum** colorpicker](https://bgrins.github.io/spectrum/) ([.js/.css](https://github.com/bgrins/spectrum)) © [_Brian Grinstead_](https://briangrinstead.com/):  [[MIT](https://github.com/bgrins/spectrum/blob/master/LICENSE)]
  jQuery: fixed deprecated functions (McBen a43f785504a73794f3deb85d9e1028a25d9132c4) (fixed in recent upstream too)

### [Leaflet](https://leafletjs.com/)-related components:
- [x] [1.4.0] [**Leaflet**](https://github.com/Leaflet/Leaflet) © [_Vladimir Agafonkin_](https://agafonkin.com/), [_CloudMade_](https://cloudmade.com/): [[LICENCE](https://github.com/Leaflet/Leaflet/blob/master/LICENSE)]
- [ ] [33ea262 + [mod](https://github.com/McBen/ingress-intel-total-conversion/commits/1901ee52534c1108402b864eff387a47a15830ca/external/leaflet.draw.js)] [**Leaflet.draw**](https://github.com/Leaflet/Leaflet.draw) © [_Jon West_](https://github.com/ddproxy), [_Jacob Toye_](https://github.com/jacobtoye), and _Leaflet_: [[MIT](https://github.com/Leaflet/Leaflet.draw/blob/develop/MIT-LICENSE.md)]
- [Shramov **leaflet-plugins**](http://psha.org.ru/b/leaflet-plugins.html) © _Pavel Shramov_, [_Bruno Bergot_](https://www.eliaz.fr/): [[MIT](https://github.com/shramov/leaflet-plugins/blob/master/LICENSE)]
  - [ ] [44935a9] [**Bing.js**](https://github.com/shramov/leaflet-plugins/blob/master/layer/tile/Bing.js)
  - [ ] [c91e49c] [**Yandex.js**](https://github.com/shramov/leaflet-plugins/blob/master/layer/tile/Yandex.js)
  - [ ] ~~[3d2d4b0] [**KML.js**](https://github.com/shramov/leaflet-plugins/blob/master/layer/vector/KML.js)~~
- [ ] [b68a264 + 1641fd5] [**toGeoJSON**](https://mapbox.github.io/togeojson/) ([togeojson.js](https://github.com/mapbox/togeojson)) © [_Mapbox_](https://www.mapbox.com/): [[BSD-2-Clause](https://github.com/mapbox/togeojson/blob/master/LICENSE)]
- [ ] [04c6d46 + [mod](https://github.com/IITC-CE/ingress-intel-total-conversion/commits/master/external/leaflet.filelayer.js)] [**Leaflet.FileLayer**](http://makinacorpus.github.io/Leaflet.FileLayer/) ([leaflet.filelayer.js](https://github.com/makinacorpus/Leaflet.FileLayer)) © [_Makina Corpus_](http://makinacorpus.com/): [[MIT](https://github.com/makinacorpus/Leaflet.FileLayer/blob/master/LICENSE)]
- [x] [0.8.0] [**Leaflet.GridLayer.GoogleMutant**](https://gitlab.com/IvanSanchez/Leaflet.GridLayer.GoogleMutant) (Leaflet.GoogleMutant.js) © [_Iván Sánchez Ortega_](https://gitlab.com/IvanSanchez): [[LICENCE](https://gitlab.com/IvanSanchez/Leaflet.GridLayer.GoogleMutant/blob/master/LICENSE)]
- [ ] [v3.4.0 + 78d77aa + 1973792] [**Leaflet.MiniMap**](https://github.com/Norkart/Leaflet-MiniMap) (Control.MiniMap.js) © [_Norkart AS_](http://www.norkart.no/): [[BSD-2-Clause](https://github.com/Norkart/Leaflet-MiniMap/blob/master/LICENSE.txt)]
- https://github.com/kartena © [_Kartena AB_](http://www.kartena.se/)
  - [ ] [7b54411] [**Leaflet.Pancontrol**](http://kartena.github.io/Leaflet.Pancontrol/) (L.Control.Pan [.js/.css/images](https://github.com/kartena/Leaflet.Pancontrol/tree/master/src)): [[BSD-2-Clause](https://github.com/kartena/Leaflet.Pancontrol/blob/master/LICENSE)]
  - [ ] [0ca1aba] [**Leaflet.zoomslider**](http://kartena.github.com/Leaflet.zoomslider/) (L.Control.Zoomslider [.js/.css](https://github.com/kartena/Leaflet.zoomslider/tree/master/src)): [[BSD-2-Clause](https://github.com/kartena/Leaflet.zoomslider/blob/master/LICENSE)]
- [[mod](https://github.com/IITC-CE/ingress-intel-total-conversion/commits/master/external/L.Geodesic.js)] [**Leaflet.Geodesic**](http://fragger.github.io/Leaflet.Geodesic/) (L.Geodesic.js) © _Kevin Brasier_: [[BSD-2-Clause](https://github.com/Fragger/Leaflet.Geodesic/blob/master/LICENSE)]
- [v0.2.6 + 1973792] [**Overlapping Marker Spiderfier**](https://github.com/jawj/OverlappingMarkerSpiderfier-Leaflet) (oms.min.js) © [_George MacKerron_](http://mackerron.com/): [[MIT](https://opensource.org/licenses/mit-license.php)]
- [7f807ae + 9077c24] [**LatLng_Bearings.js**](https://github.com/gregallensworth/Leaflet/blob/master/LatLng_Bearings.js) [[www](http://gregorthemapguy.blogspot.com/2014/02/leaflet-latlng-objects-get-bearings.html)] © _Greg Allensworth_: [No licence]

### Other:
- [ ] [6717ce2] [**autolink-js**](https://github.com/bryanwoods/autolink-js) © [_Bryan Woods_](http://bryanwoods4e.com/): [[MIT](https://github.com/bryanwoods/autolink-js/blob/master/LICENSE)]
- [ ] [f828670 + ea46405] [**delaunay.js**](https://github.com/ironwallaby/delaunay) © [_Jay LaPorte_](http://ironwallaby.github.io/): [[CC0 1.0](https://creativecommons.org/publicdomain/zero/1.0/)]
Modded (minor refactoring needed during update).
- [3acd55a] [**load.js**](https://github.com/chriso/load.js) © [_Chris O'Hara_](mailto:cohara87@gmail.com): [[MIT](https://github.com/node-js-libs/load.js/blob/master/LICENSE)]
- [5e20625 + 42ed656 + 6f8fbe6] [**s2-geometry-javascript**](https://github.com/jonatkins/s2-geometry-javascript/) (s2geometry.js) © _Jon Atkins_: [LICENSE: same as IITC]

**Comment by johnd0e** [comment]
So I've figured out exact versions of externals (as long as required modification), and now we can begin to update most of them (excl. `Leaflet.draw`, `leaflet.filelayer.js`, `togeojson.js`).

Some things still need to be clarified:
* Do we really need 522ef34f: https://github.com/IITC-CE/ingress-intel-total-conversion/pull/78#issuecomment-462323839?
Update: #122
* `Leaflet.Geodesic` has an alternative:
(https://github.com/henrythasler/Leaflet.Geodesic) (Leaflet.Geodesic.js) © [_Henry Thasler_](https://www.thasler.com/blog/): [[GPL-3.0](https://github.com/henrythasler/Leaflet.Geodesic/blob/master/LICENSE)]
* `s2geometry.js` has more advanced fork: [**s2geometry.js**](https://github.com/Daplie/s2-geometry.js/) © _Jon Atkins_, [_AJ ONeal_](https://coolaj86.com/): [[ISC](https://github.com/Daplie/s2-geometry.js/blob/master/LICENSE)]
_Moved to https://git.daplie.com/Daplie/s2-geometry.js/ (failed to connect)_
* ~~`Yandex.js` is extremely laggy~~ fixed.
~~Consider trying https://github.com/kostjerry/leaflet-yandex instead~~
* ~~`Bing.js` is extremely blurry~~ `enableRetina` fixed.
* ~`Shramov leaflet-plugins`: files `leaflet.filelayer.js`, `togeojson.js` are not present in upstream, need to search out their sources.~
There is an error in ATTRIBUTION.md, that files are not by Shramov
* [**toGeoJSON**](https://mapbox.github.io/togeojson/) ([togeojson.js](https://github.com/tmcw/togeojson)) © [_Mapbox_](https://www.mapbox.com/): [[BSD-2-Clause](https://github.com/tmcw/togeojson/blob/master/LICENSE)]
This repo is now developed at https://github.com/tmcw/togeojson: there is no major features added yet. But it is ES2015+.
* `delaunay.js` upstream is updated to more fast version, but it's still not fastest available, see performance comparison [here](https://github.com/mapbox/delaunator#performance).
  We should consider switch to https://github.com/mapbox/delaunator, it has pretty the same api, with minor differences.
  <details>

  ```
  // https://unpkg.com/delaunator@3.0.2/delaunator.js
  var triangles = tidyLinks.Delaunator.from(locations.map(function (latlng) {
    var point = map.project(latlng, tidyLinks.PROJECT_ZOOM);
    return [point.x,point.y];
  })).triangles;
  ```
  </details>

* alternative to `oms.js`: https://github.com/robertleeplummerjr/Leaflet.spiderfy/
  see https://github.com/jawj/OverlappingMarkerSpiderfier-Leaflet/pull/30, https://github.com/jawj/OverlappingMarkerSpiderfier-Leaflet/pull/31
* `load.js` is used only in `basemap-yandex.user.js` and `sync.user.js`, in form of `load(yandexApiJs).thenRun(callback);`, so I suppose it can be replaced with [`jQuery.getScript()`](https://api.jquery.com/jquery.getscript/).
BTW, there is also [jQuery.loadScript](http://marcbuils.github.io/jquery.loadscript/)
* `autolink.js` is used in single place: (and I don't feel it's necessary):
https://github.com/IITC-CE/ingress-intel-total-conversion/blob/d77562b91a5b1e0bf170b4257eb33316a0b9daa3/code/chat.js#L334
* `jquery.qrcode.min.js` is used in single place too: https://github.com/IITC-CE/ingress-intel-total-conversion/blob/d77562b91a5b1e0bf170b4257eb33316a0b9daa3/code/utils_misc.js#L207
  Perhaps it should be better to have it as plugin.
  The codebase is rather old. Alternatives: https://www.nayuki.io/page/qr-code-generator-library#javascript, https://www.cssscript.com/tag/qr-code/
* `LatLng_Bearings.js` can be replaced by gmaps geometry library function [spherical.computeHeading](https://developers.google.com/maps/documentation/javascript/reference/geometry#spherical.computeHeading).
<details><summary>
P.S.

`L.Control.Pan.css` and `L.Control.Zoomslider.css` was slightly modified, so after update we need to verify if it is still needed.

</summary>

https://github.com/iitc-project/ingress-intel-total-conversion/pull/99/:
```L.Control.Pan.css
-.leaflet-control-zoom {
+.leaflet-bar {

+/*
+  override the scale bar and push it over
+*/
+.leaflet-control-scale {
+  left: 63px !important;
+}
```
https://github.com/iitc-project/ingress-intel-total-conversion/pull/283:
```L.Control.Zoomslider.css
 .leaflet-control-zoomslider-slider {
+       background-color: #fff;
+       border-bottom: 1px solid #ccc;
```
</details>
